### PR TITLE
JAMES-2637 PUT alias route

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
+package org.apache.james.webadmin.routes;
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java]: You are missing the compulsory Apache v2 licence

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    @Override
+    public String getBasePath() {
+        return ROOT_PATH;
+    }
+
+    @Override
+    public void define(Service service) {
+        service.put(ALIAS_ADDRESS_PATH, this::throwUnknownPath);
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java]: This route is not required as the 404 default answer gives enough information IMO

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        MappingSource source = MappingSource.fromUser(User.fromLocalPartWithDomain(destinationAddress.getLocalPart(), destinationAddress.getDomain()));
+        addAlias(source, aliasSourceAddress);
+        return halt(HttpStatus.NO_CONTENT_204);
+    }
+
+    private void addAlias(MappingSource source, MailAddress aliasSourceAddress) throws RecipientRewriteTableException {
+        try {
+            recipientRewriteTable.addAliasMapping(source, aliasSourceAddress.asString());
+        } catch (MappingAlreadyExistsException e) {
+            // ignore
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java]: Why? Shouldn't this be a 500 to inform the user?

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        }
+    }
+
+    private void ensureUserDoesNotExist(MailAddress mailAddress) throws UsersRepositoryException {
+        if (usersRepository.contains(mailAddress.asString())) {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java]: This, is not as simple....

VirtualHosting tells you if local parts point to the same user or not.

 - If virtualHosting is on, then @domain is part of the username andthe full mail address is used to identify a user
 - If virtualHosting is off, then only the localpart is used to identify the user. `@bob@linagora.com` & `bob@open-paas.org` will both point to the same `bob` user.

You need to use `UsersRepository` to convert MailAddress => username

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.BAD_REQUEST_400)
+                .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+                .message("The alias source exists as an user already")
+                .haltError();
+        }
+    }
+
+    private MailAddress parseMailAddress(String address) {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java]: This piece of code is duplicated in `ForwardRoutes` & `GroupRoutes`.

Maybe we can introduce a `MailAddressParser` with a static method `parseMailAddress` that would be called in these 3 places? (a potential "addressType" parameter would help to keep the same message)

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            return new MailAddress(decodedAddress);
+        } catch (AddressException e) {
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.BAD_REQUEST_400)
+                .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+                .message("The alias is not an email address")
+                .cause(e)
+                .haltError();
+        } catch (UnsupportedEncodingException e) {
+            LOGGER.error("UTF-8 should be a valid encoding");
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java]: Why do we choose to not log and thus loose this stacktrace?

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/AliasRoutesTest.java [hunk 7]
```diff
@@ -0,0 +1,275 @@
+package org.apache.james.webadmin.routes;
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/AliasRoutesTest.java]: License

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = ALIAS_DESTINATION_ADDRESS + " or alias structure format is not valid"),
+        @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "The alias source exists as an user already"),
+        @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500,
+            message = "Internal server error - Something went bad on the server side.")
+    })
+    public HaltException addToAliasSources(Request request, Response response) throws UsersRepositoryException, RecipientRewriteTableException {
+        MailAddress aliasSourceAddress = parseMailAddress(request.params(ALIAS_SOURCE_ADDRESS));
+        ensureUserDoesNotExist(aliasSourceAddress);
+        MailAddress destinationAddress = parseMailAddress(request.params(ALIAS_DESTINATION_ADDRESS));
+        MappingSource source = MappingSource.fromUser(User.fromLocalPartWithDomain(destinationAddress.getLocalPart(), destinationAddress.getDomain()));
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java]: User.fromMailAddress would limit the boiler plate

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/AliasRoutesTest.java [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+            .build();
+    }
+
+    @AfterEach
+    void stop() {
+        webAdminServer.destroy();
+    }
+
+    @Nested
+    class NormalBehaviour {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/AliasRoutesTest.java]: Can we have tests checking that the record was added in the RRT ?

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java (hunk 3 above)
Comment (Arsnael) [reply server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java]: I think we agreed that if the alias exists already for that destination address we do nothing and still return 204. I think MappingAlreadyExistsException is thrown in that case right? So I believe it's not an issue. Wdyt?

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/AliasRoutesTest.java (hunk 9 above)
Comment (Arsnael) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/AliasRoutesTest.java]: I believe I would need the GET methods for that, that should come in a later PR?

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java (hunk 3 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/AliasRoutes.java]: I read to fast sorry.

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/AliasRoutesTest.java (hunk 9 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/AliasRoutesTest.java]: Can we get a simple test testing the underlying structure as a temporary workaround?

Shipping untested code is bad...

File: src/site/markdown/server/manage-webadmin.md [hunk 10]
```diff
@@ -1299,6 +1300,36 @@ Response codes:
...
+to be configured.
+
+Note that email addresses are restricted to ASCII character set. Mail addresses not matching this criteria will be rejected.
+
+ - [Adding a new alias to an user](#Adding_a_new_alias_to_an_user)
+
+### Adding a new alias to an user
+
+```
+curl -XPUT http://ip:port/address/aliases/user@domain.com/sources/alias@domain.com
```
Comment (blackheaven) [inline src/site/markdown/server/manage-webadmin.md]: Maybe add a body example

Comment (chibenwa) [reply src/site/markdown/server/manage-webadmin.md]: No body!

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/MailAddressParser.java [hunk 11]
```diff
@@ -19,17 +19,17 @@
 
 package org.apache.james.webadmin.routes;
 
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import javax.mail.internet.AddressException;
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/MailAddressParser.java]: You may need a line break between java & javax imports

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**